internal/controller/apiextensions/composite/environment_fetcher.go

```diff
@@ -104,37 +109,61 @@ func (f *APIEnvironmentFetcher) Fetch(ctx context.Context, cr resource.Composite
 	return env, nil
 }
 
-func (f *APIEnvironmentFetcher) fetchEnvironment(ctx context.Context, cr resource.Composite, required bool) (*Environment, error) {
-	refs := cr.GetEnvironmentConfigReferences()
-	loadedConfigs := []v1alpha1.EnvironmentConfig{}
+func getInitData(req EnvironmentFetcherRequest) (map[string]interface{}, error) {
+	if req.Revision == nil || req.Revision.Spec.Environment == nil || req.Revision.Spec.Environment.InitData == nil {
+		return map[string]interface{}{}, nil
+	}
+	return unmarshalData(req.Revision.Spec.Environment.InitData)
+}
+
+func (f *APIEnvironmentFetcher) fetchEnvironment(ctx context.Context, req EnvironmentFetcherRequest) (*Environment, error) {
+	loadedConfigs, err := f.fetchEnvironmentConfigs(ctx, req)
+	if err != nil {
+		return nil, errors.Wrap(err, errFetchEnvironmentConfigs)
+	}
+
+	mergedData, err := mergeEnvironmentData(loadedConfigs)
+	if err != nil {
+		return nil, errors.Wrap(err, errMergeData)
+	}
+	return &Environment{
+		unstructured.Unstructured{
+			Object: mergedData,
+		},
+	}, nil
+}
+
+func (f *APIEnvironmentFetcher) fetchEnvironmentConfigs(ctx context.Context, req EnvironmentFetcherRequest) ([]*v1alpha1.EnvironmentConfig, error) {
+	loadedConfigs := []*v1alpha1.EnvironmentConfig{}
+
+	// If the user provides a default environment with the composition, add it
+	// as a dummy environment config that is overwritten by all others.
+	if req.Revision != nil && req.Revision.Spec.Environment != nil && req.Revision.Spec.Environment.InitData != nil {
+		loadedConfigs = append(loadedConfigs, &v1alpha1.EnvironmentConfig{
+			Data: req.Revision.Spec.Environment.InitData,
+		})
+	}
+
+	refs := req.Composite.GetEnvironmentConfigReferences()
 	for _, ref := range refs {
-		config := v1alpha1.EnvironmentConfig{}
+		config := &v1alpha1.EnvironmentConfig{}
 		nn := types.NamespacedName{
 			Name: ref.Name,
 		}
-		err := f.kube.Get(ctx, nn, &config)
+		err := f.kube.Get(ctx, nn, config)
 		if err != nil {
 			// skip if resolution policy is optional
-			if required {
+			if req.Required {
 				return nil, errors.Wrap(err, errGetEnvironmentConfig)
 			}
 			continue
 		}
 		loadedConfigs = append(loadedConfigs, config)
 	}
-
-	mergedData, err := mergeEnvironmentData(loadedConfigs)
-	if err != nil {
-		return nil, errors.Wrap(err, errMergeData)
-	}
-	return &Environment{
-		unstructured.Unstructured{
-			Object: mergedData,
-		},
-	}, nil
+	return loadedConfigs, nil
 }
 
-func mergeEnvironmentData(configs []v1alpha1.EnvironmentConfig) (map[string]interface{}, error) {
+func mergeEnvironmentData(configs []*v1alpha1.EnvironmentConfig) (map[string]interface{}, error) {
 	merged := map[string]interface{}{}
 	for _, e := range configs {
 		if e.Data == nil {
```

Comment:
This could panic if passed a nil element, can you add a check on `e` too? or maybe just switch back everything not to use pointers altogether if there was no strong reason to do so.

Reply:
This is just a little micro-optimization: I used a pointer here to have less data copied between function calls or iterations. It is only used within the same file and there is no chance where nil pointers are added to the list.